community/modules/scripts/spack-install/scripts/install_spack_deps.yml

```diff
@@ -14,25 +14,30 @@
 
 ---
 
-- name: Install dependencies for spack installation
+- name: Install dependencies for Spack installation
   become: yes
   hosts: localhost
+  vars:
+    spack_virtualenv_path: "/usr/local/spack-python"
   tasks:
   - name: Install pip3 and git
     ansible.builtin.package:
       name:
       - python3-pip
       - git
-  - name: Gather the package facts
-    ansible.builtin.package_facts:
-      manager: auto
-  - name: Install protobuf for old releases of Python
-    when: ansible_facts.packages["python3"][0].version is version("3.7", "<") and ansible_facts.packages["python3"][0].version is version("3.5", ">=")
+    register: package
+    changed_when: package.changed
+    retries: 5
+    delay: 10
+    until: package is success
+  - name: Create virtualenv for Spack
+    # this declares our minimum support for Python 3.6+ along with ensuring the
```

Comment:
IIUC, the `declaration of 3.6 support` comes from the minimum version of pip? or something else?

Also clarifying, `dependency resolution of real-world problems` means that this version of pip will make sure protobuf is installed if needed?

Reply:
I've modified this comment for a bit more direct clarity:

- We want to support Python 3.6 because it's easy to install on common distros of Linux used in HPC (RHEL 7/8 and derivatives)
- pip 21.3.1 is the maximum pip supported by Python 3.6
- The biggest "real world" problem with pip is that the version that comes with RHEL7 (not sure about 8) is so ancient it doesn't even properly recognize package limits on Python version. So it will blindly install the most recent copy of google-cloud-storage which doesn't support Python 3.6
- The second-biggest real world problem is that before 20.3, pip would gladly break environments (including the system-wide environment), to satisfy a `pip install` command (yes, really it would do this). Now it will just fail. So, it's almost always better to create a virtual environment with minimally-scoped dependencies
- The third-biggest real world problem is that, for several days, all GCP python packages were broken when the Python release of protobuf version bumped with a breaking change to 4.21.0. TL;DR: packages should have depended on `~ 3.x` but instead had no upper bound. This unblocks `pip` to safely install the latest compatible release (3.20.3).